Templating/VOKEmbeddedTemplateExtraction.m

```diff
@@ -0,0 +1,115 @@
+//
+//  VOKEmbeddedTemplateExtraction.m
+//  Cat2Cat
+//
+//  Created by Isaac Greenspan on 8/15/15.
+//  Copyright (c) 2015 Vokal. All rights reserved.
+//
+
+#import "VOKEmbeddedTemplateExtraction.h"
+
+#import <mach-o/getsect.h>
+#import <mach-o/ldsyms.h>
+
+#import <ZipZap.h>
+
+/**
+ *  Get the content embedded into the running mach-o executable in the __c2c_tmplt_zip section of the __TEXT segment.
+ *
+ *  @return An NSData object containing the content.
+ */
+static NSData *GetEmbeddedTemplateData(void)
+{
+    // Extract the raw bytes of the __c2c_tmplt_zip (Cat2Cat's template zip archive) section of the __TEXT segment of
+    // the currently-running mach-o executable.
+    unsigned long size;
+    void *ptr = getsectiondata(&_mh_execute_header, "__TEXT",
+                               "__c2c_tmplt_zip", &size);
+    
+    // Wrap those bytes in an NSData object.
+    return [NSData dataWithBytesNoCopy:ptr
+                                length:size
+                          freeWhenDone:NO];
+}
+
+/**
+ *  Safely (relatively securely) create a temporary directory.
+ *
+ *  @return The URL of the created temporary directory.
+ */
+static NSURL *SafelyCreateTemporaryDirectory(void)
+{
+    // Set up the directory name template.
+    NSString *tempDirectoryTemplate = [NSTemporaryDirectory() stringByAppendingPathComponent:
+                                       @"Cat2Cat-Template.XXXXXX"];  // The Xs will be replaced by random characters.
+    const char *tempDirectoryTemplateCString = [tempDirectoryTemplate fileSystemRepresentation];
+    char *tempDirectoryNameCString = (char *)malloc(strlen(tempDirectoryTemplateCString) + 1);
+    strcpy(tempDirectoryNameCString, tempDirectoryTemplateCString);
+    
+    // Have the system attempt to create a directory from the template.
+    char *result = mkdtemp(tempDirectoryNameCString);
```

Comment:
From http://www.cocoawithlove.com/2009/07/temporary-files-and-folders-in-cocoa.html (and I've added that URL in the docu-comment above this function):

>To avoid concurrency issues (race conditions), you need to choose a name for any file or directory placed directly in the temporary directory that is unique but which also can't be stolen by another process between choosing the name and creating/opening the file or directory. Even though NSTemporaryDirectory() is unique for each user, you still must avoid potential conflicts within the user's account.
>
>To do this, we use the POSIX function mkstemp for files and mkdtemp for directories. The former will return an already open file descriptor for the file and the latter will have already created the directory for us.

Note that this is much less of a thing on iOS because of sandboxing—for a given iOS app, the filesystem the app sees pretty much belongs entirely and solely to the app, so there's much less concern about what a malicious app sharing the filesystem might be able to do.